Removed the object module exclusivity to define reserved class names

ErnyTech opened the pull request:
This PR removes the exclusivity of the object module of defining the
reserved class/struct names:
	- TypeInfo
	- TypeInfo_Class
	- TypeInfo_Interface
	- TypeInfo_Struct
	- TypeInfo_Pointer
	- TypeInfo_Array
	- TypeInfo_StaticArray
	- TypeInfo_AssociativeArray
	- TypeInfo_Enum
	- TypeInfo_Function
	- TypeInfo_Delegate
	- TypeInfo_Tuple
	- TypeInfo_Const
	- TypeInfo_Invariant
	- TypeInfo_Shared
	- TypeInfo_Inout
	- TypeInfo_Vector
	- Object
	- Throwable
	- Exception
	- Error
	- __cpp_type_info_ptr
	- ModuleInfo

Now you can define these classes/structs in any module, the only limit
still related to these names and the possibility to define it only in
one module.

ibuclaw commented:
Possibly you only want to remove the error first, so that the compiler doesn't recognise any of these library types outside of object.d.

I think a prerequisite to having this gambit in full would be to include some way of specially marking internally recognised types. Such as:
```
@typeinfo
class TypeInfo_Pointer
{
}
```
However the problem of having discrepancies of layout in compiler vs library still apply though. You can't just define typeinfo as whatever.

JinShil commented:
I believe the runtime type info types and other runtime types should have reserved identifiers, but I don't think the way it's currently done (restricting them to object.d) is the way to do it.

D already has two conventions for reserving identifiers:
1. Prefixing an identifier with `__`
2. Prefixing an identifer with `_d_`

The first seems to be more often used by the compiler iteself and the second seems to be the convention for identifiers in druntime, though the two seem to be often mixed up.  I, personally, don't see the need for both, but it is what it is.

If we want to polish up the compiler and runtime implementation to remove some of these inconsitencies, I recommend something like the following:
1.  Rename the `TypeInfo` classes to `_d_TypeInfo` or `__TypeInfo` to put them in the compiler's namespace
2.  Create an `alias TypeInfo = _d_TypeInfo` or `alias TypeInfo = __TypeInfo` for backward compatibility
3.  Modify the compiler to use the `_d_TypeInfo` or  `__TypeInfo` identifier
4.  Deprecate the `TypeInfo` alias
5.  Eventually remove the `TypeInfo` alias

I think that would accomplish the goal of reserving the special druntime identifiers, make the implementation consistent with the other conventions employed in the language/druntime implementation, and finally remove the object.d exclusivity restriction.

Commit:
Removed the object module exclusivity to define reserved class names

This PR removes the exclusivity of the object module of defining the
reserved class/struct names:
	- TypeInfo
	- TypeInfo_Class
	- TypeInfo_Interface
	- TypeInfo_Struct
	- TypeInfo_Pointer
	- TypeInfo_Array
	- TypeInfo_StaticArray
	- TypeInfo_AssociativeArray
	- TypeInfo_Enum
	- TypeInfo_Function
	- TypeInfo_Delegate
	- TypeInfo_Tuple
	- TypeInfo_Const
	- TypeInfo_Invariant
	- TypeInfo_Shared
	- TypeInfo_Inout
	- TypeInfo_Vector
	- Object
	- Throwable
	- Exception
	- Error
	- __cpp_type_info_ptr
	- ModuleInfo

Now you can define these classes/structs in any module, the only limit
still related to these class names is the possibility to define it only
in one module.

Signed-off-by: Ernesto Castellotti <erny.castell@gmail.com>
